File: typescript/bc-envelope/src/parameters-store.ts
import {
    type EnvelopeParameter,
    knownParameter,
    parameterName,
} from './parameter.js';

export class ParametersStore {
    readonly #dict = new Map<string, string>();

    constructor(parameters: Iterable<EnvelopeParameter> = []) {
        for (const value of parameters) {
            this.insert(value);
        }
    }

    #key(value: EnvelopeParameter): string {
        return value.kind === 'known' ? `k:${value.value}` : `n:${value.name}`;
    }

    insert(value: EnvelopeParameter): void {
        if (value.kind !== 'known') {
            throw new Error('Only known parameters can be inserted');
        }
        this.#dict.set(this.#key(value), parameterName(value));
    }

    assignedName(value: EnvelopeParameter): string | undefined {
        return this.#dict.get(this.#key(value));
    }

    name(value: EnvelopeParameter): string {
        return this.assignedName(value) ?? parameterName(value);
    }

    static nameForParameter(value: EnvelopeParameter, store?: ParametersStore): string {
        return store?.assignedName(value) ?? parameterName(value);
    }
}

export const BLANK = knownParameter(1n, '_');
export const LHS = knownParameter(2n, 'lhs');
export const RHS = knownParameter(3n, 'rhs');

export const GLOBAL_PARAMETERS = new ParametersStore([BLANK, LHS, RHS]);

File: go/bcenvelope/parameters_store.go
package bcenvelope

import "sync"

// ParametersStore maps parameters to their assigned names.
type ParametersStore struct {
	dict map[uint64]parameterEntry // known parameters keyed by value
}

type parameterEntry struct {
	parameter Parameter
	name      string
}

// NewParametersStore creates a new store populated with the provided parameters.
// Only known (numeric) parameters can be inserted.
func NewParametersStore(parameters ...Parameter) *ParametersStore {
	store := &ParametersStore{
		dict: make(map[uint64]parameterEntry),
	}
	for _, p := range parameters {
		store.Insert(p)
	}
	return store
}

// Insert adds a known parameter into the store.
// Panics if a named parameter is inserted.
func (s *ParametersStore) Insert(parameter Parameter) {
	if parameter.kind != parameterKindKnown {
		panic("only known parameters can be inserted into ParametersStore")
	}
	name := parameter.Name()
	s.dict[parameter.value] = parameterEntry{parameter: parameter, name: name}
}

// AssignedName returns the assigned name for a parameter if it exists in the store.
func (s *ParametersStore) AssignedName(parameter Parameter) (string, bool) {
	if s == nil {
		return "", false
	}
	if parameter.kind == parameterKindKnown {
		entry, ok := s.dict[parameter.value]
		if !ok {
			return "", false
		}
		return entry.name, true
	}
	return "", false
}

// Name returns the name for a parameter from the store, falling back to the parameter's own name.
func (s *ParametersStore) Name(parameter Parameter) string {
	if name, ok := s.AssignedName(parameter); ok {
		return name
	}
	return parameter.Name()
}

// NameForParameter returns the name of a parameter using an optional store.
func NameForParameter(parameter Parameter, store *ParametersStore) string {
	if store != nil {
		if name, ok := store.AssignedName(parameter); ok {
			return name
		}
	}
	return parameter.Name()
}

// Clone returns an independent copy of the store.
func (s *ParametersStore) Clone() *ParametersStore {
	if s == nil {
		return NewParametersStore()
	}
	clone := NewParametersStore()
	for _, entry := range s.dict {
		clone.Insert(entry.parameter)
	}
	return clone
}

// Global parameters store singleton.
var (
	globalParameters     *ParametersStore
	globalParametersOnce sync.Once
)

// GlobalParameters returns the lazily initialized global parameters store.
func GlobalParameters() *ParametersStore {
	globalParametersOnce.Do(func() {
		globalParameters = NewParametersStore(
			ParameterBlank, ParameterLHS, ParameterRHS,
		)
	})
	return globalParameters
}
